Add GitHub MCP microservice for repository queries

DhruvK278 opened the pull request:
Introduces a new GitHub MCP microservice with FastAPI for handling repository metadata queries. Adds client, server, and service modules for the MCP, updates the intent analysis prompt and toolkit to route repo_support queries and to the new service, and implements the repository_query  tool. Also updates Discord cog lifecycle methods and bumps aio-pika and aiohttp dependencies.

# feat: Add GitHub MCP service

Closes #117

## 📝 Description
This PR adds a new GitHub MCP (Microservice Control Plane) service that allows the agent to query GitHub repository metadata (stars, forks, open issues, etc.) in a structured way.

The service is implemented using FastAPI and is integrated with the MCP server-client pattern. It securely fetches data using a GitHub Personal Access Token (PAT) stored in environment variables.

---

## 🔧 Changes Made

This PR introduces a new **GitHub MCP microservice** with FastAPI for handling repository metadata queries.  
The following changes were made:

- Added `github_mcp_service.py` → service layer with GitHub API calls.
- Added `github_mcp_server.py` → FastAPI MCP server exposing endpoints.
- Integrated MCP into the toolkit so that repo-related queries are routed correctly.
- Updated intent analysis prompt to support `repo_support` queries.
- Implemented `repository_query` tool for agent use.
- Updated Discord cog lifecycle methods for MCP integration.
- Bumped `aio-pika` and `aiohttp` dependencies for improved stability.

## 📷 Screenshots or Visual Changes (if applicable)

Repo_query
<img width="1268" height="847" alt="image" src="https://github.com/user-attachments/assets/84643089-9ad1-4ca7-8a85-a481f324d020" />

<img width="1057" height="805" alt="image" src="https://github.com/user-attachments/assets/e7ca1739-251f-4eae-a1d1-1177b20bf059" />


## 🤝 Collaboration

Guidance under @smokeyScraper 

### ✅ Checklist

- [ ] I have read the contributing guidelines.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if applicable).
- [ ] Any dependent changes have been merged and published in downstream modules.



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added GitHub support: ask about repos/orgs, list repos/issues, and get repo/org summaries.
  - Added a lightweight GitHub MCP server with health check and a runnable startup script.

- **Refactor**
  - More robust intent classification and JSON handling for GitHub queries; prompts now prefer GitHub data over web search.
  - Discord startup moved to cog-based startup for token cleanup.

- **Chores**
  - Bumped Python requirement to 3.10 and added new dependencies.
  - Added GITHUB_ORG to example environment config.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

DhruvK278 committed:
Add GitHub MCP microservice for repository queries

Introduces a new GitHub MCP microservice with FastAPI for handling repository metadata queries. Adds client, server, and service modules for the MCP, updates the intent analysis prompt and toolkit to route repo_support queries to the new service, and implements the repository_query tool. Also updates Discord cog lifecycle methods and bumps aio-pika and aiohttp dependencies.

DhruvK278 commented:
@coderabbitai review

DhruvK278 committed:
Delete backend/app/agents/devrel/github/README.md

DhruvK278 committed:
added coderabbit changes

DhruvK278 commented:
@coderabbit review

DhruvK278 commented on backend/app/agents/devrel/github/services/github_mcp_server.py:
```diff
@@ -0,0 +1,96 @@
+        logger.exception("Error in MCP endpoint")
+        return {"error": str(e)}
+
+@app.post("/repo_info")
+async def get_repo_info(request: RepoInfoRequest):
+
+    try:
+        if not github_service:
+            raise HTTPException(status_code=500, detail="GitHub service not available")
+        
+        result = github_service.repo_query(request.owner, request.repo)
+        
+        if "error" in result:
+            return RepoInfoResponse(status="error", data={}, error=result["error"])
+        
+        return RepoInfoResponse(status="success", data=result)
+        
+    except Exception as e:
+        logger.error(f"Error getting repo info: {e}")
+        raise HTTPException(status_code=500, detail=str(e))
+
```
your current fix blocks the /repo_info to perform action leading to error 502
 RuntimeWarning: coroutine 'GitHubMCPService.repo_query' was never awaited
 
 will keep using the same code

DhruvK278 commented:
@smokeyScraper Please have a look with my fixed changes :)

DhruvK278 commented:
@coderabbitai review

smokeyScraper commented on backend/app/agents/devrel/github/prompts/intent_analysis.py:
```diff
@@ -1,9 +1,9 @@
-- web_search: Search the web for information  
-- contributor_recommendation: Finding the right people to review PRs, assign issues, or collaborate (supports both issue URLs and general queries)
-- repo_support: Questions about codebase structure, dependencies, impact analysis, architecture
+- web_search: Search the web for general information not available through GitHub API
```
No need to specify "not available through GitHub API", this might lead the agent to use GitHub API first, maybe, and might lead to hallucinations. It's best if it decides on its own.

smokeyScraper commented on backend/app/agents/devrel/github/prompts/intent_analysis.py:
```diff
@@ -1,9 +1,9 @@
-- repo_support: Questions about codebase structure, dependencies, impact analysis, architecture
+- web_search: Search the web for general information not available through GitHub API
+- contributor_recommendation: Finding the right people to review PRs, assign issues, or collaborate
+- repo_support: Questions about repository information, structure, stats, issues, stars, forks, description, or any repository metadata
```
could you please rename this to somewhat like GitHub support? ```repo_support``` would be using Falkor DB and will be very specific to codebase-related queries.

smokeyScraper reviewed (changes requested):
I was a bit constrained over to the part that users do not need to specify the organization, and we'll be passing the organization name in the .env itself, as one deployed agent will work for an org only, right?
could you please align it in this way? so, yeah, it'll help simply answer the queries like,
"What all repositories does this organization have?"

The rest of everything seems fine to me. Response too is great!
Thanks a lot for your contribution @DhruvK278 !

PS: Please do once consider all the coderrabit suggestions, a few of them seem important to me.

DhruvK278 committed:
Refactor repo info tool to github_support

Replaces the old repository_query tool with github_support for handling repository metadata queries. Updates prompt classification, toolkit logic, MCP client/server endpoints, and introduces github_support.py with improved naming and separation of concerns.

DhruvK278 committed:
Refactor GitHub support to use org from env and add org repo listing

GitHub support tools and services now default to using the organization from environment variables when only a repo name is provided. Added endpoints and logic for listing organization repositories and stats, improved intent classification for repo-related queries, and updated prompts to clarify GitHub toolkit usage. Updated env.example to document new environment variables.

DhruvK278 commented:
@coderabbitai review

DhruvK278 committed:
Added CoderabbitAI changes

DhruvK278 commented:
@coderabbitai review

DhruvK278 commented on backend/app/agents/devrel/github/tools/github_support.py:
```diff
@@ -0,0 +1,110 @@
+        # Fallback org if none detected
+        org = org or DEFAULT_ORG
+
+        # --- Top repos ---
+        if "top" in q and "repo" in q:
+            repos = await github_mcp_service.get_org_repositories(org)
+            repos = sorted(repos, key=lambda r: r["stars"], reverse=True)[:10]
+            return {
+                "status": "success",
+                "message": f"Top repositories for {org}",
+                "repositories": repos,
+            }
```
if isinstance(repos, dict) and "error" in repos:
                  ^^^^^
UnboundLocalError: cannot access local variable 'repos' where it is not associated with a value

DhruvK278 committed:
Improve error handling for GitHub org repo listing (Coderabbit fix)

Changed HTTP status code to 503 when GitHub service is unavailable and made repo listing run in a separate thread for better performance. Enhanced error handling in github_support.py to return error details if repository fetching fails and made repo sorting more robust.

DhruvK278 commented:
@coderabbitai review

smokeyScraper commented on backend/app/agents/devrel/prompts/react_prompt.py:
```diff
@@ -13,10 +13,14 @@
 
 AVAILABLE ACTIONS:
-1. web_search - Search the web for external information
+1. web_search - Search the web for external information (ONLY if you cannot find info in GitHub or FAQs)
```
Please try not to make anything very specific. This might make the agent think on preferring GitHub/FAQs first.

smokeyScraper commented on backend/app/agents/devrel/github/github_toolkit.py:
````diff
@@ -55,7 +68,25 @@ async def classify_intent(self, user_query: str) -> Dict[str, Any]:
 
             import json
-            result = json.loads(response.content.strip())
+            import re
+
+            content = response.content.strip()
+
+            candidates = []
+            cb = re.search(r'```(?:json)?\s*({[\s\S]*?})\s*```', content, flags=re.IGNORECASE)
+            if cb:
+                candidates.append(cb.group(1))
+            candidates.extend(m.group(0) for m in re.finditer(r'\{[\s\S]*?\}', content))
+
+            result = None
+            for payload in candidates:
+                try:
+                    result = json.loads(payload)
+                    break
+                except json.JSONDecodeError:
+                    continue
+            if result is None:
+                raise json.JSONDecodeError("No valid JSON object found in LLM response", content, 0)
````
can you please elaborate on what this regex is for? seems to me like for extracting repo/org from the user query.
But won't it work without regex as used specifically?

A bit confused cuz github_support.py too has regex defined.

smokeyScraper reviewed (changes requested):
LGTM!
A few clarifications on the extensive use of Regex part and reviewing the coderabbit suggestion would be enough!

Thanks a lot!

DhruvK278 committed:
Added changes suggested by mentor

minimized regex use in github_toolkit.py and some changes in react_prompt

smokeyScraper reviewed (approved):
Thanks a lot @DhruvK278!!
This took a long way... Thanks for staying in..
